[Sema] Diagnose instant deallocation of instances assigned to weak/unowned variables

hamishknight opened the pull request:
When performing a binding/assignment to a `weak` or `unowned` variable/property from an initialiser call, emit a warning that the instance will be immediately deallocated.

For example:
```swift
class C {}

// warning: instance will be immediately deallocated as 'c' is a 'weak' variable
weak var c = C()
```
Before requesting a review, I'd like to ask how seriously we need to consider code like this:
```swift
class C {
  init() {
    someGlobal = self
  }
}

var someGlobal: C?

// warning: instance will be immediately deallocated as 'c' is a 'weak' variable
weak var c = C()
```

With the implementation in this PR, we'll still diagnose the binding even though the instance will be retained by the global variable. How should this be handled? IMO we should still diagnose as initialisers really shouldn't have side effects like this.

We *could* limit the diagnostic to "simple" initialisers, i.e those that we 1) know the body of and 2) can prove that `self` never escapes. However IMO this would be too limiting, especially considering we would likely have to discount initialisers that call instance methods. I think the usefulness of applying the diagnostic in the general case far outweighs the very small number of corner cases where it would be a false positive.

I did consider changing the warning message to account for this corner case (something like "instance will _likely_ be immediately deallocated ..."), but I think that just adds confusion in the general case, so decided against it.

Finally, it's worth noting that this PR doesn't handle tuple shuffles, for example:
```swift
class C {}

weak var (a, b): (a: C?, b: C?) = (a: C(), b: C()) // warning

weak var (a, b): (b: C?, a: C?) = (a: C(), b: C()) // no warning
```

I didn't feel it was worth the implementation complexity to support them, but am happy to do so (either in this PR or a follow-up) if it's felt that support would be beneficial.

Resolves [SR-7095](https://bugs.swift.org/browse/SR-7095).

hamishknight committed:
[Sema] Fix implementation of getValueProvidingExpr()

We should trying to find the value providing expr(s) from the semantic providing expr.

hamishknight committed:
[Sema] Diagnose immediate deallocation of instances assigned to non-owning variables

When performing a binding/assignment to a weak or unowned variable/property from an initialiser call, emit a warning that the instance will be immediately deallocated.

CodaFi commented:
I wonder if this would be easier to do in Definite Initialization checking.

hamishknight commented:
@CodaFi Good idea, I hadn't considered that! I'll take a look into that.

hamishknight commented:
@CodaFi Looking at it, I don't believe it can done as a part of the Definite Initialization pass, as that only runs on `MarkUninitializedInst` instructions.

It could probably be written as a separate guaranteed SIL pass, but it would still require knowledge of the AST as it needs to check for calls to constructors (unless I'm mistaken, I don't think there's an easy way to check if a given SIL `function_ref` refers to a constructor). Though if we want to keep the ability to look through things like force unwraps, optional bindings and tuple elements, I don't think the implementation of the diagnostic at SIL level will be any simpler than in Sema.

Therefore I'll leave it as a Sema diagnostic for now.

hamishknight commented:
cc. @jrose-apple, any thoughts on this diagnostic?

jrose-apple commented:
I'm leery of there being no way to silence it, because the "escaping in initializer" pattern really does happen (though not often). I don't think that means it needs to be limited, just that there needs to be a way to say "yes, I know what's going on" that's more concise than using a temporary variable.

@rjmccall, any thoughts?

hamishknight commented:
@jrose-apple Okay, in that case should the diagnostic be "instance will *likely* be immediately deallocated ..." along with the option to silence?

Also how about asking the user to explicitly refer to the initialiser as e.g `C.init()` in order to silence the warning?

jrose-apple commented:
I'm okay with the diagnostic text as is, as long as there's a note that says "do X to silence this warning". I don't like any of the options but the parentheses seem least bad to me. :-/

gparker42 commented:
If false positives really are rare then requiring a temporary variable seems like an acceptable cost.

jrose-apple commented:
Maybe that's fair. What do other people think?

CodaFi commented:
The same warning is implemented in Clang in exactly this situation

```objective-c
__weak NUIColor *color = [[NUIColor alloc] init];
// Assigning retained object to weak variable; object will be released after assignment
```

Coincidentally, I also think this warning text is the model that should be used for this patch.

hamishknight commented:
The Clang diagnostic is nice in that it it's accurate for both the general case and corner cases where the instance escapes (as it talks in terms of retaining and releasing rather than saying the instance will be deallocated).

However I don't think we should expect users to know what retaining and releasing is; these concepts don't appear to be mentioned in either [the official language guide](https://developer.apple.com/library/content/documentation/Swift/Conceptual/Swift_Programming_Language/index.html#//apple_ref/doc/uid/TP40014097-CH3-ID0) or existing user-facing diagnostics. The [ARC section](https://developer.apple.com/library/content/documentation/Swift/Conceptual/Swift_Programming_Language/AutomaticReferenceCounting.html#//apple_ref/doc/uid/TP40014097-CH20-ID48) of the language guide just talks in terms of strong/weak/unowned references to instances, allocation, initialization, deinitialization and deallocation.

jrose-apple commented:
I agree, "retaining" and "releasing" are not common Swift terms. Let's take this, then, but please add a test in definite_init_diagnostics.swift that shows that assigning to a strong variable first silences the diagnostic…

…hang on, you can't just assign to a strong variable first. It would have to be a strong variable *in a nested scope.*

Maybe we don't care, though.

hamishknight commented:
@jrose-apple Added tests for that (and one for assigning through an immediately evaluated closure, e.g `= { C() }()`).

xwu commented:
Can I propose a wording? Here's the relevant TSPL text:

> A weak reference is a reference that does not keep a strong hold on the instance it refers to, and so does not stop ARC from disposing of the referenced instance.

> Because a weak reference does not keep a strong hold on the instance it refers to, it’s possible for that instance to be deallocated while the weak reference is still referring to it.

So, maybe, "assigning instance to a 'weak' variable; instance will be immediately deallocated without a strong reference"?

hamishknight commented:
@xwu I like the additional bit about requiring a strong reference, though IMO it would be nice to also get the variable name in the diagnostic (especially given we also emit a note showing you where it's declared).

So maybe "assigning instance to the 'weak' variable 'someVariable'; instance will be immediately deallocated without a strong reference"? Though I don't know whether that's getting a bit too long for  single warning.

Alternatively, we could the bit about requiring a strong reference as a separate note, so:

- warning: "instance will be immediately deallocated as 'someVariable' is a 'weak' variable"
- note: "there is no strong reference preventing the instance from being deallocated"
- note: "'someVariable' declared here"

Thoughts?

hamishknight commented:
@xwu Was the thumbs up in favour of the extended warning, or adding the additional note? Though actually I think a better phrasing for that note would be "a strong reference is required to prevent the instance from being deallocated".

xwu commented:
Ha, sorry, that was ambiguous. I like the revised note actually. Didn’t think it needs the second note about where the variable is declared, as that’s the location where the warning comes up?

hamishknight commented:
@xwu The location of the declaration can differ from the location of the warning if it's an assignment rather than a binding, for example:
```swift
class C {}

// note: 'someVariable' declared here
weak var someVariable: C?
// ...

// warning: instance will be immediately deallocated as 'someVariable' is a 'weak' variable
// note: a strong reference is required to prevent the instance from being deallocated
someVariable = C()
```

We could omit the 'declared here' note on bindings (i.e `weak var c = C()`), but I don't think there's any harm in leaving it in.

hamishknight commented:
Just added the additional note; I think this PR is ready for review now.

hamishknight commented:
@jrose-apple Would it be possible to get this reviewed in time for 4.2? I appreciate that it's additive and therefore low priority, so no worries if not.

jrose-apple commented on lib/Sema/MiscDiagnostics.cpp:
```diff
@@ -1997,6 +1997,177 @@ bool swift::diagnoseArgumentLabelError(TypeChecker &TC, const Expr *expr,
+  // This is to ensure we don't diagnose on constructors such as
+  // Optional.init(nilLiteral:).
+  auto selfType = constructorDecl->getDeclContext()->getSelfTypeInContext();
+  if (!selfType->allowsOwnership())
```
You already have the type; it's the type of `initExpr`.

hamishknight replied:
Not necessarily for failable initialisers, as they'll have an optional type (which doesn't allow ownership even though we want to diagnose on them).

jrose-apple replied:
Uh, hm. This still seems roundabout to me but I guess it's okay.

jrose-apple commented on lib/Sema/MiscDiagnostics.cpp:
```diff
@@ -1997,6 +1997,177 @@ bool swift::diagnoseArgumentLabelError(TypeChecker &TC, const Expr *expr,
+              diag::unowned_assignment_requires_strong)
+    .highlight(initExpr->getSourceRange());
+
+  TC.diagnose(varDecl->getStartLoc(), diag::decl_declared_here,
```
This should just use `varDecl` as the location, which will point to the name of the declaration and also handle variables imported from other modules.

jrose-apple commented on lib/Sema/MiscDiagnostics.cpp:
```diff
@@ -1997,6 +1997,177 @@ bool swift::diagnoseArgumentLabelError(TypeChecker &TC, const Expr *expr,
+  if (varDecl->getDeclContext()->isTypeContext())
+    storageKind = SK_Property;
+
+  TC.diagnose(initExpr->getStartLoc(),
```
Ideally these notes would point to the assignment, not the constructor call.

hamishknight replied:
@jrose-apple Thanks for the review!

When you say the assignment, do you mean highlighting the entire `AssignExpr`? For example:
```swift
weak var c4: C?
c4 = C()
^~~~~~~~
```

I agree that would be better, though I don't believe we can the same for arbitrary bindings, for example here are the ranges we currently highlight for a tuple pattern:

```swift
class C {}
weak var (c1, c2, c3) = (C(), C(), C())
                         ^~~  ^~~  ^~~
```
It would be ambiguous which element of the tuple we'd be referring to if we highlighted the entire thing instead.

Would it be worth special-casing a 'basic' named pattern, such that we highlight like this:

```swift
weak var c4 = C()
         ^~~~~~~~
```
instead of just on `C()`?

jrose-apple replied:
No, I mean the location of the caret:

```swift
weak var c4: C?
c4 = C()
   ^ ~~~
```

That's something you can do without messing with the highlight ranges.

hamishknight replied:
@jrose-apple Ah right, I see! We can certainly do that for assignments, though for bindings it doesn't look like we keep the source loc of the equals token(s) about on the `PatternBindingDecl`. I can look into implementing that if you want, though in the interest of getting this PR moving, I think it would be best to do so in a follow up PR. Thoughts?

jrose-apple replied:
Sure, that sounds fine. And yes, keeping around a SourceLoc in PatternBindingEntry for the equal sign sounds like the right thing to do.

jrose-apple commented on lib/Sema/MiscDiagnostics.cpp:
```diff
@@ -1997,6 +1997,177 @@ bool swift::diagnoseArgumentLabelError(TypeChecker &TC, const Expr *expr,
+      for (unsigned i = 0, e = TP->getNumElements(); i != e; ++i) {
+        TuplePatternElt &elt = TP->getElement(i);
+        Pattern *subPattern = elt.getPattern();
+        Expr *subInitExpr = TE->getElement(i);
```
You should `lookThroughExprsToImmediateDeallocation` again here, no?

hamishknight replied:
@jrose-apple We don't have to; if it's a nested `TuplePattern`, then we `lookThroughExprsToImmediateDeallocation` on its `initExpr` (in the recursive call) before digging into the elements, and if it's a nested `NamedPattern` then `diagnoseUnownedImmediateDeallocationImpl` already calls it.

jrose-apple replied:
Ah, I missed that this was a recursive call and not directly calling the `Impl` variant. My mistake!

jrose-apple commented on lib/Sema/MiscDiagnostics.h:
```diff
@@ -64,6 +65,18 @@ bool diagnoseArgumentLabelError(TypeChecker &TC, const Expr *expr,
+/// If \c destExpr refers to a declaration with a non-owning attribute, such as
+/// 'weak' or 'unowned' and \c initializer refers to a class constructor,
+/// emit a warning that the assigned instance will be immediately deallocated.
+void diagnoseUnownedImmediateDeallocation(Expr *destExpr, Expr *initializer,
```
Nitpick: More `const`s on unmodified AST node parameters, please!

Nitpick: Use `\p` for parameters in doc comments rather than `\c`.

(here and below)

jrose-apple reviewed (commented):
Heh. I was hoping either @gparker42 or @CodaFi would weigh back in on the workaround being terrible if you actually did mean to have no extra strong references for a local besides a constructor escape:

```swift
weak var ref: Foo?
do {
  let strongRef = Foo()
  ref = strongRef
}
```

But since this is so rare maybe it's not an issue. It's more likely that it's fine for the strong reference to last until the end of the function and have a weak property after that.

I have a handful of small implementation comments as well.

hamishknight committed:
Merge remote-tracking branch 'upstream/master' into warn-weak-insta-dealloc

hamishknight committed:
[Sema] Revise #14875 in response to feedback

- Diagnose on the location of the '=' equals sign for assignments
- Use the variable decl as the location for the note
- Make AST node params 'const'

hamishknight commented:
@jrose-apple Thanks again for the feedback! I've gone ahead and pushed those changes.

It's worth noting that the user can also silence the warning by using an immediately executed closure, e.g:

```swift
weak var c = { C() }()
```

Though I doubt that many people will actually think to do that (and those that do won't find it particularly attractive).

jrose-apple reviewed (approved):
Looks good to me. Thanks, Hamish!

jrose-apple commented:
@swift-ci Please test source compatibility
